components/higher-order/with-notices/index.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import uuid from 'uuid/v4';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, createHigherOrderComponent } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import NoticeList from '../../notice/list';
+
+/**
+ * Override the default edit UI to include notices if supported.
+ *
+ * @param  {function|Component} OriginalComponent Original component.
+ * @return {Component}                            Wrapped component.
+ */
+export default createHigherOrderComponent( ( OriginalComponent ) => {
+	return class WrappedBlockEdit extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.createNotice = this.createNotice.bind( this );
+			this.createErrorNotice = this.createErrorNotice.bind( this );
+			this.removeNotice = this.removeNotice.bind( this );
+			this.removeAllNotices = this.removeAllNotices.bind( this );
+
+			this.state = {
+				noticeList: [],
+			};
+
+			this.noticeOperations = {
+				createNotice: this.createNotice,
+				createErrorNotice: this.createErrorNotice,
+				removeAllNotices: this.removeAllNotices,
+				removeNotice: this.removeNotice,
+			};
+		}
+
+		/**
+		* Function passed down as a prop that adds a new notice.
+		*
+		* @param {Object} notice  Notice to add.
+		*/
+		createNotice( notice ) {
+			const noticeToAdd = notice.id ? notice : { ...notice, id: uuid() };
+			this.setState( ( state ) => ( {
+				noticeList: [ ...state.noticeList, noticeToAdd ],
+			} ) );
+		}
+
+		/**
+		* Function passed as a prop that adds a new error notice.
+		*
+		* @param {string} msg  Error message of the notice.
+		*/
+		createErrorNotice( msg ) {
+			this.createNotice( { status: 'error', content: msg } );
+		}
+
+		/**
+		* Removes a notice by id.
+		*
+		* @param {string} id  Id of the notice to remove.
+		*/
+		removeNotice( id ) {
+			this.setState( ( state ) => ( {
+				noticeList: state.noticeList.filter( ( notice ) => notice.id !== id ),
+			} ) );
+		}
+
+		/**
+		* Removes all notices
+		*/
+		removeAllNotices() {
+			this.setState( {
+				noticeList: [],
+			} );
+		}
+
+		render() {
+			return (
+				<OriginalComponent
+					noticeList={ this.state.noticeList }
+					noticeOperations={ this.noticeOperations }
+					noticeUI={
```

Comment:
This seems nicer! Wondering if it could make it clear things up even more if `noticeOperations` are properties on `noticeUI`? Do you think that would be strange?

Reply:
Related with https://github.com/WordPress/gutenberg/pull/4203#discussion_r188730846 although we pass a noticeUI I wanted it to be a totally optional prop and allow components to ignore it if they just want the notices without relying on the default UI we provide.